### DISCUSSION TITLE
More/Better typing?

### DISCUSSION THREAD

**Discussion by junpenglao** [blackjax-devs/blackjax#555] [New features]
TL;dr: Should we introduce [jaxtyping](https://github.com/google/jaxtyping) and [beartype](https://github.com/beartype/beartype)?

# Currently Problem
As noted in https://github.com/blackjax-devs/blackjax/pull/543#issue-1745581275, I am not happy about the way `logdensity` is annotated (e.g., see [example](https://github.com/blackjax-devs/blackjax/blob/cf94b27d6a87d17b64a030d7a73d87227fb1080b/blackjax/mcmc/hmc.py#LL40C21-L40C21)). While they are Array (as they are in most cases should be output of a Jax function), we annotate them as float to empathizes they should be scalar.

# Potential Solution
While [jaxtyping](https://github.com/google/jaxtyping) is introduced to solve this, it would make the code much longer 
```python
# Accepts floating-point 2D arrays with matching dimensions
def matrix_multiply(x: Float[Array, "dim1 dim2"],
                    y: Float[Array, "dim2 dim3"]
                  ) -> Float[Array, "dim1 dim3"]:
    ...
```
We also have a lot of function that accepts nested structure (PyTree), although we can introduce [beartype](https://github.com/beartype/beartype) so runtime check speed would not be a concern.

**Comment by albcab** [comment]
I like runtime type-checking, it's useful for efficiently writing bug-free code and a great complement to testing. It seems like we'd need to change various `float` and `bool` to `Float/Bool[Array, "1"]` and we can get runtime type-checking? Or am I missing something else? If that is the case then it sounds like a good trade-off.
